File: src/modules/theme_manager.rs
use crate::{args::ThemeManagerArgs, get_image_type};
use colored::Colorize;
use std::env;
use std::path::Path;
use std::process::Command;
use std::{thread, time};

fn is_root() -> bool {
    unsafe { libc::geteuid() == 0 }
}

fn wait() {
    let timeout = time::Duration::from_secs(25);
    let start = time::Instant::now();

    loop {
        if env::var("DISPLAY").is_ok() {
            println!("DISPLAY is set! Running..");
            return; // Success, exit normally
        }

        if start.elapsed() >= timeout {
            eprintln!("Error: DISPLAY is not set after waiting 25 seconds.");
            std::process::exit(1); // Exit with error
        }

        thread::sleep(time::Duration::from_secs(1)); // Check once per second
    }
}

fn set_gnome_theme(args: &ThemeManagerArgs) {
    let image_type = get_image_type();

    if image_type.contains("arcturus") {
        // Set wallpaper for dark theme
        run_command(
            "gsettings",
            format!(
                "set org.gnome.desktop.background picture-uri-dark {}",
                &args.wallpaper
            )
            .as_str(),
            "GNOME wallpaper",
        );

        // Set wallpaper for light theme
        run_command(
            "gsettings",
            format!(
                "set org.gnome.desktop.background picture-uri {}",
                &args.wallpaper
            )
            .as_str(),
            "GNOME wallpaper",
        );

        // Set cursor theme
        run_command(
            "gsettings",
            format!(
                "set org.gnome.desktop.interface cursor-theme {}",
                &args.cursor
            )
            .as_str(),
            "Set Cursor theme",
        );

        // Set icon theme
        run_command(
            "gsettings",
            format!("set org.gnome.desktop.interface icon-theme {}", &args.icons).as_str(),
            "GNOME icons",
        );

        // Set gtk theme
        run_command(
            "gsettings",
            format!("set org.gnome.desktop.interface gtk-theme {}", &args.theme).as_str(),
            "GNOME theme",
        );

        // Apply blur-my-shell settings
        run_command(
            "dconf",
            "load -f /org/gnome/shell/extensions/blur-my-shell/ < /usr/share/horizon/configs/arcturus/blur-my-shell.dconf",
            "Blur my shell settings",
        );

        // Apply blur-my-shell settings
        run_command(
            "dconf",
            "load -f /org/gnome/shell/extensions/dash-to-dock/ < /usr/share/horizon/configs/arcturus/dash-to-dock.dconf",
            "Blur my shell settings",
        );

        // Enable Blur my shell
        run_command(
            "gnome-extensions",
            "enable blur-my-shell@aunetx",
            "Enable blur my shell",
        );

        // Enable dash to dock
        run_command(
            "gnome-extensions",
            "enable dash-to-dock@micxgx.gmail.com",
            "Enable dash to dock",
        );
    }
}

fn set_plasma_theme(args: &ThemeManagerArgs) {
    let image_name = get_image_type();

    if image_name.contains("plasma") || image_name.contains("aster") {
        // Apply the theme
        run_command(
            "plasma-apply-colorscheme",
            &args.theme,
            "Plasma colorscheme",
        );

        // Apply the cursor theme
        run_command("plasma-apply-cursortheme", &args.cursor, "Cursor theme");

        // Apply the icon theme
        run_command("/usr/libexec/plasma-changeicons", &args.icons, "Icon theme");

        // Apply the wallpaper
        run_command("plasma-apply-wallpaperimage", &args.wallpaper, "Wallpaper");
    } else {
        println!("{}", "Unsupported desktop environment.".red());
    }
}

fn run_command(command: &str, arg: &str, description: &str) {
    let full_command = format!("{} {}", command, arg);

    let status = Command::new("/bin/sh")
        .args(["-c", full_command.as_str()])
        .status()
        .expect("Failed to run the command");

    if !status.success() {
        eprintln!("{}: Failed to set the {}", "ERR".bold().red(), description);
    }
}

fn disable_systemd_service() {
    let _systemd_service_path: Option<&Path>;
    let service_name: &str;

    // Determine which systemd service exists
    if Path::new("/usr/share/horizon").exists() {
        _systemd_service_path = Some(Path::new(
            "/usr/lib/systemd/user/horizon-theme-manager.service",
        ));
        service_name = "horizon-theme-manager.service";
    } else if Path::new("/usr/share/nova").exists() {
        _systemd_service_path = Some(Path::new(
            "/usr/lib/systemd/user/nova-theme-manager.service",
        ));
        service_name = "nova-theme-manager.service";
    } else if Path::new("/usr/share/umbra").exists() {
        _systemd_service_path = Some(Path::new(
            "/usr/lib/systemd/user/umbra-theme-manager.service",
        ));
        service_name = "umbra-theme-manager.service";
    } else {
        panic!("{}: No systemd service found!", "ERR".bold().red());
    };

    // Disable the systemd service using systemctl
    let status = Command::new("systemctl")
        .args(["--user", "disable", service_name])
        .status()
        .expect("Failed to run systemctl disable command");

    if status.success() {
        println!(
            "{}: Successfully disabled {}",
            "DONE".bold().green(),
            service_name
        );
    } else {
        eprintln!("{}: Failed to disable {}", "ERR".bold().red(), service_name);
    }
}

pub fn theme_manager(args: &ThemeManagerArgs) {
    let image_type = get_image_type();

    if is_root() {
        println!(
            "{}: Please don't run the theme manager as root.",
            "WARN".yellow()
        );
        return;
    } else {
        wait();

        if image_type.contains("arcturus") {
            set_gnome_theme(&args);
        } else if image_type.contains("aster") {
            set_plasma_theme(&args);
        }

        disable_systemd_service();
    }
}

File: modules/themeManager.go
package modules

import (
	"bufio"
	"fmt"
	"os"
	"os/exec"
	"path/filepath"
	"strings"
	"time"

	"github.com/darkmatterorg/nebula/config"
	"github.com/darkmatterorg/orbit/utils"
)

func wait() {
	timeout := 25 * time.Second
	deadline := time.Now().Add(timeout)

	for {
		if _, exists := os.LookupEnv("DISPLAY"); exists {
			utils.Notice("DISPLAY is set.. Running")
			return
		}

		if time.Now().After(deadline) {
			utils.Error("DISPLAY isn't set even after 25 seconds, exiting.")
			os.Exit(1)
		}

		time.Sleep(1 * time.Second) // Check once per second
	}
}

func Setplasmatheme(theme string, icons string, cursor string, wallpaper string) {
	if utils.IsProcessRunning("gnome-shell") {
		if config.Config.Insults {
			utils.Error("Are you dumb? Don't run the Plasma theme functions on non-Plasma environment's")
		} else {
			utils.Error("Don't run Plasma theme functions on non-Plasma environment's")
		}
		return
	}

	wait()

	plasma_colorscheme := exec.Command("plasma-apply-colorscheme", theme)
	plasma_colorscheme_err := plasma_colorscheme.Run()

	if plasma_colorscheme_err != nil {
		utils.Error("Failed to set the Plasma colorscheme")
	}

	plasma_icons := exec.Command("/usr/libexec/plasma-changeicons", icons)
	plasma_icons_err := plasma_icons.Run()

	if plasma_icons_err != nil {
		utils.Error("Failed to set the Icon theme")
	}

	plasma_cursor := exec.Command("plasma-apply-cursortheme", cursor)
	plasma_cursor_err := plasma_cursor.Run()

	if plasma_cursor_err != nil {
		utils.Error("Failed to set the cursor theme")
	}

	plasma_wallpaper := exec.Command("plasma-apply-wallpaperimage", wallpaper)
	plasma_wallpaper_err := plasma_wallpaper.Run()

	if plasma_wallpaper_err != nil {
		utils.Error("Failed to set the wallpaper.")
	}
}

func enableGnomeExtensions(filePath string) {
	// Open the file
	file, err := os.Open(filePath)
	if err != nil {
		message := fmt.Sprintf("Failed to open file: %v\n", err)
		utils.Error(message)
		return
	}
	defer file.Close()

	// Read the file line by line
	scanner := bufio.NewScanner(file)
	for scanner.Scan() {
		extension := strings.TrimSpace(scanner.Text()) // Remove whitespace
		if extension == "" {
			continue // Skip empty lines
		}

		// Enable the extension
		cmd := exec.Command("gnome-extensions", "enable", extension)
		err := cmd.Run()
		if err != nil {

			utils.Error(fmt.Sprintf("Failed to enable extension %s\n", extension))

		} else {
			utils.Done(fmt.Sprintf("Enabled extension: %s\n", extension))
		}
	}

	// Check for errors during scanning
	if err := scanner.Err(); err != nil {
		utils.Error(fmt.Sprintf("Error reading file: %v\n", err))
	}
}

func loadDconfFiles(directory string) error {
	// Read all files in the directory
	files, err := os.ReadDir(directory)
	if err != nil {
		return fmt.Errorf("failed to read directory: %v", err)
	}

	// Loop through the files in the directory
	for _, file := range files {
		// Only process regular files (not directories)
		if !file.IsDir() {
			filePath := filepath.Join(directory, file.Name())
			// Execute the dconf command to load each file
			cmd := exec.Command("dconf", "load", "/org/gnome/", filePath)

			// Run the command and get output
			output, err := cmd.CombinedOutput()
			if err != nil {
				utils.Error(fmt.Sprintf("loading file %s: %v\n", file.Name(), err))
				fmt.Println("Output:", string(output))
				continue
			}

			utils.Done(fmt.Sprintf("Successfully loaded %s\n", file.Name()))
		}
	}

	return nil
}

func Setgnometheme(icons string, wallpaper_dark string, wallpaper_light string, cursor string, theme string) {
	if utils.IsProcessRunning("plasmashell") {
		if config.Config.Insults {
			utils.Error("Are you dumb? Don't run GNOME theme functions on non-GNOME environment's")
		} else {
			utils.Error("Don't run GNOME theme functions on non-GNOME environment's")
		}
		return
	}

	wait()

	// Set wallpaper for dark theme
	wallpaper_darkmode := exec.Command("gsettings", "set", "org.gnome.desktop.background", "picture-uri-dark", wallpaper_dark)

	wallpaper_darkmode_err := wallpaper_darkmode.Run()

	if wallpaper_darkmode_err != nil {
		utils.Error("Failed to set the dark mode wallpaper!")
	}

	// Set wallpaper for light theme
	wallpaper_lightmode := exec.Command("gsettings", "set", "org.gnome.desktop.abckground", "picture-uri", wallpaper_light)

	wallpaper_lightmode_err := wallpaper_lightmode.Run()

	if wallpaper_lightmode_err != nil {
		utils.Error("Failed to set the light mode wallpaper!")
	}

	set_cursor := exec.Command("gsettings", "set", "org.gnome.desktop.interface", "cursor-theme", cursor)

	set_cursor_err := set_cursor.Run()

	if set_cursor_err != nil {
		utils.Error("Failed to set the cursor theme!")
	}

	icon_theme := exec.Command("gsettings", "set", "org.gnome.desktop.interface", "icon-theme", icons)

	icon_theme_err := icon_theme.Run()

	if icon_theme_err != nil {
		utils.Error("Failed to set the icon theme!")
	}

	set_gtk_theme := exec.Command("gsettings", "set", "org.gnome.desktop.interface", "gtk-theme", theme)

	set_gtk_theme_err := set_gtk_theme.Run()

	if set_gtk_theme_err != nil {
		utils.Error("Failed to set the GTK theme!")
	}

	switch {
	case strings.Contains(image_type, "arcturus"):
		enableGnomeExtensions("/usr/share/horizon/configs/arcturus/gnome_extensions.txt")
		loadDconfFiles("/usr/share/horizon/configs/arcturus")
	case strings.Contains(image_type, "umbra"):
		enableGnomeExtensions("/usr/share/umbra/configs/gnome_extensions.txt")
	}
}
